kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt

```diff
@@ -56,20 +56,76 @@ class DateProperties(private val date: Date) {
 }
 
 class DateFunctions(private val date: Date) {
+    val addYears = function(
+        command = { n: nat ->
+            val nextYear = date.year + n
+            val nextMonth = date.month
+            val nextDay = min(date.day, daysInMonth(nextYear, nextMonth))
+
+            mk_Date(nextYear, nextMonth, nextDay)
+        },
+        post = { n, result -> result == date.functions.addMonths(n * MonthsPerYear) }
+    )
 
-    val addMonths: (integer) -> Date = function(
+    val subtractYears = function(
+        command = { n: nat ->
+            val nextYear = date.year - n
+            val nextMonth = date.month
+            val nextDay = min(date.day, daysInMonth(nextYear, nextMonth))
+
+            mk_Date(nextYear, nextMonth, nextDay)
+        },
+        pre = { n -> date.year > n },
+        post = { n, result -> result == date.functions.subtractMonths(n * MonthsPerYear) }
+    )
+
+    val addMonths: (nat) -> Date = function(
         command = { n ->
-            val nextMonth = (date.month.safeToInt() + n - 1).mod(MonthsPerYear.safeToInt()).toNat() + 1u
-            val yearShift = (date.month.safeToInt() + n - 1).floorDiv(MonthsPerYear.safeToInt())
-            val nextYear = (date.year.safeToInt() + yearShift).toNat()
+            val yearsToAdd = n / MonthsPerYear
+            val monthsToAdd = n % MonthsPerYear
+            val willOverflow = MonthsPerYear < date.month + monthsToAdd
+
+            val nextYear = if (willOverflow) {
+                date.year + (yearsToAdd + 1u)
+            } else {
+                date.year + yearsToAdd
+            }
```

Comment:
:pick: it's probably much of a muchness here but fyi you could do 
```suggestion
            val (nextYear, nextMonth) = if (willOverflow) {
                mk_(date.year + (yearsToAdd + 1u), date.month + (monthsToAdd - MonthsPerYear))
            } else {
                mk_(date.year + yearsToAdd, date.month + monthsToAdd)
            }
```

Reply:
I didn't realise I could do this!  That's been rolled up now, it's a lot cleaner.